docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,718 @@
+    # kubeAPIServer:
+    #   featureGates:
+    #     SomeKubernetesFeature: true
+    #   runtimeConfig:
+    #     scheduling.k8s.io/v1alpha1: true
+    #   oidcConfig:
+    #     caBundle: |
+    #       -----BEGIN CERTIFICATE-----
+    #       Li4u
+    #       -----END CERTIFICATE-----
+    #     clientID: client-id
+    #     groupsClaim: groups-claim
+    #     groupsPrefix: groups-prefix
+    #     issuerURL: https://identity.example.com
+    #     usernameClaim: username-claim
+    #     usernamePrefix: username-prefix
+    #     signingAlgs: RS256,some-other-algorithm
+    #-#-# only usable with Kubernetes >= 1.11
+    #     requiredClaims:
+    #       key: value
+    #   admissionPlugins:
+    #   - name: PodNodeSelector
+    #     config: |
+    #       podNodeSelectorPluginConfig:
+    #         clusterDefaultNodeSelector: <node-selectors-labels>
+    #         namespace1: <node-selectors-labels>
+    #         namespace2: <node-selectors-labels>
+    #   auditConfig:
+    #     auditPolicy:
+    #       configMapRef:
+    #         name: auditpolicy
+    # kubeControllerManager:
+    #   featureGates:
+    #     SomeKubernetesFeature: true
+    #   horizontalPodAutoscaler:
+    #     syncPeriod: 30s
+    #     tolerance: 0.1
+    #-#-# only usable with Kubernetes < 1.12
+    #     downscaleDelay: 15m0s
+    #     upscaleDelay: 1m0s
+    #-#-# only usable with Kubernetes >= 1.12
+    #     downscaleStabilization: 5m0s
+    #     initialReadinessDelay: 30s
+    #     cpuInitializationPeriod: 5m0s
+    # kubeScheduler:
+    #   featureGates:
+    #     SomeKubernetesFeature: true
+    # clusterAutoscaler:
+    #   scaleDownUtilizationThreshold: 0.5
+    #   scaleDownUnneededTime: 30m
+    #   scaleDownDelayAfterAdd: 60m
+    #   scaleDownDelayAfterFailure: 10m
+    #   scaleDownDelayAfterDelete: 10s
+    #   scanInterval: 10s
+    worker:
+    - name: cpu-worker
+      minimum: 3
+      maximum: 5
+    # maxSurge: 1
+    # maxUnavailable: 0
+      machine:
+        type: m5.large
+        image:
+          name: <some-os-name>
+          version: <some-os-version>
+        # providerConfig:
+        #   <some-os-specific-configuration>
+      volume:
+        type: gp2
+        size: 20Gi
+    # providerConfig:
+    #   <some-provider-specific-worker-config>
+    # labels:
+    #   key: value
+    # annotations:
+    #   key: value
+    # taints: # See also https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
+    # - key: foo
+    #   value: bar
+    #   effect: NoSchedule
+    # kubernetes:
+    #   kubeProxy:
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+    #     mode: IPVS
+    #   kubelet:
+    #     cpuCFSQuota: true
+    #     cpuManagerPolicy: none
+    #     podPidsLimit: 10
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+      zones:
+      - europe-central-1a
```

Comment:
> black magic.

Not really, they are just distributed evenly between the zones (if uneven then the first pool will get one more). Straight-forward IMHO.

The problem with your approach is that the user has to define worker pools multiple times, and especially their configuration (like kubelet settings, max surge, and many more) need to be duplicated many times. This will lead to an explosion of the `worker` section which is not necessary from my point of view. As said above, it's very easy to understand how the distribution works.

Reply:
I would keep the kubelet settings the same, but rather move there only the amount of workers:

```yaml
    workes:
    - name: cpu-worker
      zones:
      - name: europe-central-1a
        minimum: 3
        maximum: 5
      # maxSurge: 1
      # maxUnavailable: 0
      - name: europe-central-1b
        minimum: 1
        maximum: 3
      machine:
        type: m5.large
        image:
          name: <some-os-name>
          version: <some-os-version>
        # providerConfig:
        #   <some-os-specific-configuration>
      volume:
        type: gp2
        size: 20Gi
    # providerConfig:
    #   <some-provider-specific-worker-config>
    # labels:
    #   key: value
    # annotations:
    #   key: value
    # taints: # See also https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
    # - key: foo
    #   value: bar
    #   effect: NoSchedule
    # kubernetes:
    #   kubelet:
    #     cpuCFSQuota: true
    #     cpuManagerPolicy: none
    #     podPidsLimit: 10
    #     featureGates:
    #       SomeKubernetesFeature: true
```

With this is clearly specified that I want machines with a specific count.

One thing which is not clear is how to define where a worker is going to be placed in - for example in AWS you can put the workers in a specific subnet (which is a availability zone-scoped resource).
I don't see how this can be specified for this API